src/btree/row_srch.c

```diff
@@ -298,25 +319,41 @@ restart:	page = current->page;
 		 * If we end up somewhere other than the last slot, it's not a
 		 * right-side descent.
 		 */
-		if (pindex->entries != base - 1)
+		if (pindex->entries != base)
 			descend_right = false;
 
+		/*
+		 * If on the last slot (the key is larger than any key on the
+		 * page), check for an internal page split race.
+		 */
+		if (pindex->entries == base) {
+append:			if (parent_pindex != NULL &&
+			    __wt_split_intl_race(
+			    session, current->home, parent_pindex)) {
+				if ((ret = __wt_page_release(
+				    session, current, 0)) != 0)
+					return (ret);
+
+				skiplow = skiphigh = 0;
+				goto restart_root;
+			}
+		}
+
 descend:	/*
 		 * Swap the current page for the child page. If the page splits
 		 * while we're retrieving it, restart the search in the current
 		 * page; otherwise return on error, the swap call ensures we're
 		 * holding nothing on failure.
 		 */
-		switch (ret = __wt_page_swap(session, current, descent, 0)) {
-		case 0:
+		if ((ret = __wt_page_swap(session, current, descent, 0)) == 0) {
 			current = descent;
-			break;
-		case WT_RESTART:
+			continue;
+		}
+		if (ret == WT_RESTART) {
 			skiphigh = skiplow = 0;
-			goto restart;
-		default:
-			return (ret);
+			goto restart_page;
```

Comment:
@keithbostic I'm not convinced it's good enough to restart on the same page here. Could you convince me? The particular scenario I'm cautious about is if the ``__wt_page_swap`` call triggers a forced eviction, which in turn triggers a split deepen (or a cascade of them all the way up to the root, and then a deepen).